cdap-security/src/main/java/co/cask/cdap/security/store/FileSecureStore.java

```diff
@@ -0,0 +1,463 @@
+ * If that is successful then the temporary file is renamed atomically to the secure store file.
+ * If anything fails during this process then the keystore reverts to the last successfully written file.
+ * The keystore is flushed to the filesystem after every put and delete.
+ */
+class FileSecureStore implements SecureStore, SecureStoreManager {
+  private static final Logger LOG = LoggerFactory.getLogger(FileSecureStore.class);
+
+  private static final String SCHEME_NAME = "jceks";
+  private static final Gson GSON = new Gson();
+
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+  private final KeyStore keyStore;
+
+  FileSecureStore(CConfiguration cConf) throws IOException {
+    // Get the path to the keystore file
+    String pathString = cConf.get(Constants.Security.Store.FILE_PATH);
+    Path dir = Paths.get(pathString);
+    path = dir.resolve(cConf.get(Constants.Security.Store.FILE_NAME));
+
+    // Get the keystore password
+    password = cConf.get(Constants.Security.Store.FILE_PASSWORD).toCharArray();
+
+    keyStore = locateKeystore(path, password);
+    ReadWriteLock lock = new ReentrantReadWriteLock(true);
+    readLock = lock.readLock();
+    writeLock = lock.writeLock();
+  }
+
+  /**
+   * Stores an element in the secure store.
+   * @param name Name of the element to store
+   * @param data The data that needs to be securely stored
+   * @param description User provided description of the entry.
+   *@param properties Metadata associated with the data  @throws IOException
+   */
+  @Override
+  public void put(String name, byte[] data, String description, Map<String, String> properties) throws IOException {
+    SecureStoreMetadata meta = SecureStoreMetadata.of(name, description, properties);
+    SecureStoreData secureStoreData = new SecureStoreData(meta, data);
+    Key key = null;
+    writeLock.lock();
+    try {
+      if (keyStore.containsAlias(name)) {
+        key = keyStore.getKey(name, password);
+      }
+      keyStore.setKeyEntry(name, new KeyStoreEntry(secureStoreData, meta), password, null);
+      // Attempt to persist the store.
+      flush();
+    } catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableKeyException e) {
+      // We failed to store the key in the key store. Throw an IOException.
+      throw new IOException("Failed to store the key. ", e);
+    } catch (IOException ioe) {
+      // The key was stored in the keystore successfully but we failed to flush it to the file system.
+      // Remove the key from the keystore and throw an IOException.
+      try {
+        if (key == null) {
+          deleteFromStore(name, password);
+        } else {
+          keyStore.setKeyEntry(name, key, password, null);
+        }
+      } catch (KeyStoreException | UnrecoverableKeyException | NoSuchAlgorithmException e) {
+        LOG.error("Failed to recover the store after a failed flush.");
```

Comment:
The exception will not make it clear to the caller what the underlying error was. We are throwing IOException with the keystore exception added as suppressed so the actual error will be hidden from the client.